### PR TITLE
perf: datavzrd wrapper `v3.12.1`, offer-excel configurable, free disk space for CI, dynamic sleuth_init mem_mb, pure download rules as localrules

### PR DESCRIPTION
**Pull request by dlaehnemann**
this then uses `datavzrd=2.36.10`

### CODE REVIEW

File: .test/config/config.yaml [hunk 1]
```diff
@@ -93,6 +93,9 @@ enrichment:
     # the species specified by resources -> ref -> species above
     pathway_database: "panther"
 
+report:
+  offer_excel: true
```
Comment (fxwiegand) [inline .test/config/config.yaml:97]:
```suggestion
  offer_excel: false
```

Depending on the size of the input data this might cause extremely long runtimes as the excel writer in rust requires to read all of the input data into memory before one can write it out:

https://github.com/datavzrd/datavzrd/blob/556feb7b6a39745b42900d27fbcad26ee050e158/src/render/portable/mod.rs#L1944-L1981

Comment (dlaehnemann) [reply .test/config/config.yaml:97]: Many thanks for checking the template and for the clarification!

Did you already check for better options for xlsx file writing? Have you for example seen this one:
https://docs.rs/simple-xlsx-writer/latest/simple_xlsx_writer/

At least it doesn't read everything into memory. But not sure, if it will be any quicker and if the interface works for how you parse stuff in datavzrd...

Comment (fxwiegand) [reply .test/config/config.yaml:97]: This is literally the one we are using right now 😄

Comment (dlaehnemann) [reply .test/config/config.yaml:97]: Hmmm. Maybe switch to something else, then... :thinking:

But yeah, I think this switch of the default here is good enough for now. I'll add a comment warning of the performance hit when activating this...

Comment (fxwiegand) [reply .test/config/config.yaml:97]: > Hmmm. Maybe switch to something else, then... 🤔

I am not sure there is something that would allow something like flushing the writer every thousand records for example. But it'd great to have something like that for sure. Might even have to do with excel itself maybe?

> But yeah, I think this switch of the default here is good enough for now. I'll add a comment warning of the performance hit when activating this...

Yeah a warning when manually activated sounds good!

**Comment by dlaehnemann** [comment]
For now, I'll turn off the reactivated testing of isoformSwitchAnalyzer, to get the tests to pass here and merge this PR.

And then, I'll fix isoformSwitchAnalyzer on a new branch, where we'll also fix the Ensembl cDNA fasta file, as the current failure is due to the Ensembl GTF and the Ensembl cDNA FASTA file being out of sync, see here:
https://www.r-bloggers.com/2020/01/comparing-ensembl-gtf-and-cdna/

Removing Ensembl cDNA FASTA redundancies (the same HGNC gene symbol appearing with multiple Ensembl `gene_id`s due to scaffolds and other issues), will probably remove most of the FASTA-only transcript IDs as well, so should be part of the isoformSwitchAnalyzer fix...

File: workflow/resources/datavzrd/diffexp-template.yaml [hunk 2]
```diff
@@ -1,8 +1,10 @@
+__use_yte__: true
+
 name: ?f"Differential expression analysis for model {wildcards.model}"
 datasets:
   genes_representative:
     path: ?input.genes_representative
-    offer-excel: true
+    offer-excel: ?params.offer_excel in ["true", "True"]
```
Comment (johanneskoester) [inline workflow/resources/datavzrd/diffexp-template.yaml:7]: I don't like the check for true and True as strings. The yaml schema should ensure that the value in the config yaml is a boolean. And then, there is no need to have these additional checks here.

File: workflow/rules/common.smk [hunk 3]
```diff
@@ -360,7 +356,7 @@ def all_input(wildcards):
         )
     )
 
-    if is_activated(config["diffsplice"]):
+    if config["diffsplice"]["activate"]:
```
Comment (johanneskoester) [inline workflow/rules/common.smk:359]: Why this change?

File: workflow/rules/datavzrd.smk [hunk 4]
```diff
@@ -64,13 +14,21 @@ rule spia_datavzrd:
...
+        "v3.11.0/utils/datavzrd"
 
 
 rule diffexp_datavzrd:
     input:
-        config="results/datavzrd/diffexp/{model}.yaml",
+        # the config file may be a yte template, with access to input, params and wildcards
+        # analogous to Snakemake's generic template support:
+        # https://snakemake.readthedocs.io/en/stable/snakefiles/rules.html#template-rendering-integration
+        # For template processing, __use_yte__: true has to be stated in the config file
```
Comment (johanneskoester) [inline workflow/rules/datavzrd.smk:29]:
```suggestion
```

File: workflow/rules/datavzrd.smk [hunk 5]
```diff
@@ -85,19 +43,22 @@ rule diffexp_datavzrd:
...
             labels={"model": "{model}"},
         ),
-    params:
-        model=get_model,
     log:
         "logs/datavzrd-report/diffexp.{model}/diffexp.{model}.log",
+    params:
+        extra="",
+        model=get_model,
+        offer_excel=lookup(within=config, dpath="report/offer_excel", default="false"),
```
Comment (johanneskoester) [inline workflow/rules/datavzrd.smk:51]: Default should be a boolean.

File: workflow/rules/datavzrd.smk [hunk 6]
```diff
@@ -116,5 +77,8 @@ rule go_enrichment_datavzrd:
         ),
     log:
         "logs/datavzrd-report/go_enrichment-{model}/go_enrichment-{model}_{gene_fdr}.go_term_fdr_{go_term_fdr}.log",
+    params:
+        offer_excel=lookup(within=config, dpath="report/offer_excel", default="false"),
```
Comment (johanneskoester) [inline workflow/rules/datavzrd.smk:81]:
```suggestion
        offer_excel=lookup(within=config, dpath="report/offer_excel", default=False),
```

File: workflow/rules/common.smk [hunk 7]
```diff
@@ -80,10 +80,6 @@ def check_config():
 check_config()
 
 
-def is_activated(config_element):
-    return config_element["activate"] in {"true", "True"}
-
-
 def get_model(wildcards):
```
Comment (johanneskoester) [inline workflow/rules/common.smk:83]: It was always not good that we allowed strings here. The value of activate has to be boolean and we should ensure that via the schema.

**Review by johanneskoester — CHANGES_REQUESTED** [review]
Epic, thanks a lot! Just some minor things below.

**Comment by fxwiegand** [comment]
It would probably also makes sense to upgrade to 2.36.15 as this will probably save loads of space in the resulting reports 😊

File: workflow/rules/common.smk (hunk 3 above)
Comment (dlaehnemann) [reply workflow/rules/common.smk:359]: The `is_activated()` function was not working as intended (any more). Nowadays, the yaml seems to be parsing settings of `true` directly into a boolean, and that boolean was failing the string-based comparison. Also, this was the only place that function was still used. So instead of fixing the function, I removed it -- one layer less of abstraction.

File: workflow/resources/datavzrd/diffexp-template.yaml (hunk 2 above)
Comment (dlaehnemann) [reply workflow/resources/datavzrd/diffexp-template.yaml:7]: Yeah, that's a good idea. Actually, I have no clue why the `true` or `false` in this case got parsed into a string, while in the diffsplice case it seems to have turned into a boolean. Neither of them was mentioned in the `config.schema.yaml` so far...

File: workflow/rules/datavzrd.smk (hunk 6 above)
Comment (dlaehnemann) [reply workflow/rules/datavzrd.smk]: Already done in response to previous comment.

File: workflow/rules/datavzrd.smk (hunk 5 above)
Comment (dlaehnemann) [reply workflow/rules/datavzrd.smk]: Already done in response to previous comment.

File: workflow/rules/common.smk (hunk 7 above)
Comment (dlaehnemann) [reply workflow/rules/common.smk:83]: Already done in response to previous comment.

File: workflow/rules/common.smk (hunk 3 above)
Comment (johanneskoester) [reply workflow/rules/common.smk:359]: ok
